src/types/rule.cpp

```diff
@@ -332,18 +386,40 @@ void Rule::removeString(const std::string& id)
  */
 void Rule::addTag(const std::string& tag)
 {
-	_tags.push_back(tag);
+	//find iterator behind tags in TokenStream
+	TokenIt end = ++_tags.back();
+	TokenIt newTagIt = _tokenStream->insert(end, TokenType::TAG, Literal(tag));
+	_tags.push_back(newTagIt);
 }
 
 /**
  * Removes tag from the rule.
  *
  * @param tag Tag to remove.
+ * @return true iff there was corresponding tag and it was removed
  */
 void Rule::removeTags(const std::string& tag)
 {
-	auto new_end = std::remove(_tags.begin(), _tags.end(), tag);
-	_tags.erase(new_end, _tags.end());
+	for (auto it = _tags.begin(); it != _tags.end(); ++it)
```

Comment:
I would put the body of this for-loop inside curly braces to make it obvious that this multi-line condition belongs to it. I am fan of for-loops without curly braces for their bodies only in case of one-line bodies.